File: apps/ios/Optio/Core/UI/Detail.swift
import SwiftUI

/// Standard header for detail screens (Task, Job run, Review, Agent, Session,
/// Local terminal): a material block with the badge, a title-less state line
/// (`Running · 3m 12s · claude-sonnet`) and a single accent row only when the
/// thing needs you.
struct DetailHeader<Accessory: View>: View {
    let state: String
    var tone: Tone? = nil
    /// `·`-joined facts: duration, model, repo, branch. Mono parts via `Text.mono`.
    var line: Text? = nil
    /// Second line (mono path, PR link text).
    var secondary: Text? = nil
    /// Accent row shown only when the item needs the user.
    var needsYou: String? = nil
    /// Show the shared Claude usage pill (`AccountUsagePill`) at the end of the
    /// second line — the number that decides whether another session can start.
    var showsUsage = false
    @ViewBuilder var accessory: Accessory

    init(state: String, tone: Tone? = nil, line: Text? = nil, secondary: Text? = nil, needsYou: String? = nil,
         showsUsage: Bool = false, @ViewBuilder accessory: () -> Accessory = { EmptyView() }) {
        self.state = state
        self.tone = tone
        self.line = line
        self.secondary = secondary
        self.needsYou = needsYou
        self.showsUsage = showsUsage
        self.accessory = accessory()
    }

    private var resolvedTone: Tone { tone ?? Tone.forState(state) }

    private func secondaryText(_ text: Text) -> some View {
        text.font(.monoFootnote).foregroundStyle(.secondary).lineLimit(1).truncationMode(.head)
    }

    var body: some View {
        VStack(alignment: .leading, spacing: Spacing.s) {
            HStack(spacing: Spacing.s) {
                StatusBadge(text: state, tone: resolvedTone)
                if let line {
                    line.font(.subheadline).foregroundStyle(.secondary).lineLimit(1).truncationMode(.middle)
                }
                Spacer(minLength: 0)
                accessory
            }
            if let secondary, showsUsage {
                // Path and pill share a line when both fit; otherwise the pill takes its own.
                ViewThatFits(in: .horizontal) {
                    HStack(spacing: Spacing.s) {
                        secondaryText(secondary)
                        Spacer(minLength: 0)
                        AccountUsagePill().fixedSize()
                    }
                    VStack(alignment: .leading, spacing: Spacing.xs) {
                        secondaryText(secondary)
                        AccountUsagePill()
                    }
                }
            } else if let secondary {
                secondaryText(secondary)
            } else if showsUsage {
                AccountUsagePill()
            }
            if let needsYou {
                HStack(spacing: 6) {
                    StateDot(tone: .accent)
                    Text(needsYou).font(.footnote.weight(.medium)).foregroundStyle(Tone.accent.textStyle)
                }
            }
        }
        .frame(maxWidth: .infinity, alignment: .leading)
        .padding(.horizontal, Spacing.l)
        .padding(.vertical, Spacing.m)
        .background(.regularMaterial)
        .overlay(alignment: .bottom) { Divider() }
        .modifier(UsageObserverIf(enabled: showsUsage))
    }
}

/// `observesUsage()` only for headers that show the pill.
private struct UsageObserverIf: ViewModifier {
    let enabled: Bool
    func body(content: Content) -> some View {
        if enabled { content.observesUsage() } else { content }
    }
}

/// One message bubble: user = trailing on a secondary fill, agent = leading plain prose.
struct MessageBubble: View {
    enum Role { case user, agent, system }
    let role: Role
    let text: String
    var meta: String? = nil
    var pending = false

    var body: some View {
        VStack(alignment: role == .user ? .trailing : .leading, spacing: 3) {
            Text(LocalizedStringKey(text))
                .font(role == .system ? .footnote : .body)
                .foregroundStyle(role == .system ? AnyShapeStyle(.secondary) : AnyShapeStyle(.primary))
                .textSelection(.enabled)
                .padding(.horizontal, role == .user ? Spacing.m : 0)
                .padding(.vertical, role == .user ? Spacing.s : 0)
                .background {
                    if role == .user {
                        Radius.bubbleShape.fill(.fill.secondary)
                    }
                }
                .frame(maxWidth: 320, alignment: role == .user ? .trailing : .leading)
            if let meta {
                Text(meta).font(.caption2).foregroundStyle(pending ? AnyShapeStyle(AppTheme.accent) : AnyShapeStyle(.tertiary))
            }
        }
        .frame(maxWidth: .infinity, alignment: role == .user ? .trailing : .leading)
    }
}

/// Monochrome pipeline strip: done = primary, current = accent, future = quaternary.
struct PipelineStrip: View {
    let steps: [String]
    let current: Int
    var failed = false

    var body: some View {
        HStack(spacing: Spacing.xs) {
            ForEach(Array(steps.enumerated()), id: \.offset) { i, step in
                VStack(alignment: .leading, spacing: Spacing.xs) {
                    Capsule().fill(fill(i)).frame(height: 3)
                    Text(step).font(.caption2).foregroundStyle(i <= current ? AnyShapeStyle(.primary) : AnyShapeStyle(.tertiary)).lineLimit(1)
                }
            }
        }
    }

    private func fill(_ i: Int) -> AnyShapeStyle {
        if i < current { return AnyShapeStyle(.primary) }
        if i == current { return AnyShapeStyle(failed ? Color.red : AppTheme.accent) }
        return AnyShapeStyle(.quaternary)
    }
}

/// Segmented section switcher inside a detail screen (Logs · Activity · …).
struct DetailTabs<T: Hashable>: View {
    let options: [(T, String)]
    @Binding var selection: T

    var body: some View {
        Picker("Section", selection: $selection) {
            ForEach(options, id: \.0) { value, label in Text(label).tag(value) }
        }
        .pickerStyle(.segmented)
        .padding(.horizontal, Spacing.l)
        .padding(.vertical, Spacing.s)
        .sensoryFeedback(.selection, trigger: selection)
    }
}

File: apps/android/core/ui/src/main/kotlin/dev/optio/core/ui/components/Detail.kt
package dev.optio.core.ui.components

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.ExperimentalLayoutApi
import androidx.compose.foundation.layout.FlowRow
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.RowScope
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.widthIn
import androidx.compose.foundation.text.selection.SelectionContainer
import androidx.compose.material3.LocalContentColor
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.CompositionLocalProvider
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.testTag
import androidx.compose.ui.text.AnnotatedString
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import dev.optio.core.ui.theme.OptioTheme
import dev.optio.core.ui.theme.Radius
import dev.optio.core.ui.theme.Spacing
import dev.optio.core.ui.theme.Tone
import dev.optio.core.ui.theme.medium
import dev.optio.core.ui.usage.AccountUsagePill
import dev.optio.core.ui.usage.ObservesUsage

/**
 * The standard header of detail screens (iOS `DetailHeader`: Task, Job run, Review, Agent,
 * Session, Local terminal): the [state] badge and a title-less `·`-joined [line] (`Running · 3m 12s
 * · Sonnet`), an optional mono [secondary] line (path, branch, PR), and a single accent row only
 * when the thing [needsYou]. [accessory] sits at the end of the badge row (a PR link button).
 *
 * [lineMaxLines] lets a long [line] wrap under itself instead of losing its middle (one line by
 * default, cut in the middle like iOS; a middle cut can split a value such as `$0.0…1k`).
 *
 * [secondaryTruncation] cuts a [secondary] line that doesn't fit: at the start by default, so a
 * path keeps its leaf; free text such as a description reads better cut at the end
 * ([Truncation.END]).
 *
 * [showsUsage] adds the shared Claude usage pill ([AccountUsagePill]) after the secondary line (on
 * the same line when both fit) and keeps the usage poller running while the header is on screen
 * ([ObservesUsage]). Both need `LocalUsageStore`; without it the pill is simply absent.
 */
@OptIn(ExperimentalLayoutApi::class)
@Composable
fun DetailHeader(
    state: String,
    modifier: Modifier = Modifier,
    tone: Tone? = null,
    line: AnnotatedString? = null,
    lineMaxLines: Int = 1,
    secondary: AnnotatedString? = null,
    secondaryTruncation: Truncation = Truncation.HEAD,
    needsYou: String? = null,
    showsUsage: Boolean = false,
    accessory: @Composable RowScope.() -> Unit = {},
) {
    val colors = OptioTheme.colors
    val type = OptioTheme.type
    if (showsUsage) ObservesUsage()
    Column(
        modifier = modifier
            .fillMaxWidth()
            .background(MaterialTheme.colorScheme.surfaceContainerLow)
            .testTag("detail-header"),
    ) {
        Column(
            Modifier.fillMaxWidth().padding(horizontal = Spacing.l, vertical = Spacing.m),
            verticalArrangement = Arrangement.spacedBy(Spacing.s),
        ) {
            Row(verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(Spacing.s)) {
                StatusBadge(text = state, tone = tone ?: Tone.forState(state))
                Box(Modifier.weight(1f)) {
                    if (line != null) {
                        Text(
                            line,
                            style = type.subheadline,
                            color = colors.secondaryLabel,
                            maxLines = lineMaxLines,
                            overflow = if (lineMaxLines == 1) TextOverflow.MiddleEllipsis else TextOverflow.Ellipsis,
                        )
                    }
                }
                CompositionLocalProvider(LocalContentColor provides colors.secondaryLabel) { accessory() }
            }
            when {
                secondary != null && showsUsage -> FlowRow(
                    modifier = Modifier.fillMaxWidth(),
                    horizontalArrangement = Arrangement.SpaceBetween,
                    verticalArrangement = Arrangement.spacedBy(Spacing.xs),
                    itemVerticalAlignment = Alignment.CenterVertically,
                ) {
                    SecondaryLine(secondary, secondaryTruncation, Modifier.padding(end = Spacing.s))
                    AccountUsagePill()
                }
                secondary != null -> SecondaryLine(secondary, secondaryTruncation)
                showsUsage -> AccountUsagePill()
            }
            if (needsYou != null) {
                Row(verticalAlignment = Alignment.CenterVertically, horizontalArrangement = Arrangement.spacedBy(6.dp)) {
                    StateDot(Tone.ACCENT)
                    Text(needsYou, style = type.footnote.medium(), color = Tone.ACCENT.textColor)
                }
            }
        }
        InsetDivider(start = 0.dp)
    }
}

@Composable
private fun SecondaryLine(text: AnnotatedString, truncation: Truncation, modifier: Modifier = Modifier) {
    Text(
        text,
        style = OptioTheme.type.monoFootnote,
        color = OptioTheme.colors.secondaryLabel,
        maxLines = 1,
        overflow = truncation.overflow,
        modifier = modifier,
    )
}

/** Who wrote a [MessageBubble]. */
enum class MessageRole { USER, AGENT, SYSTEM }

/**
 * One chat message (iOS `MessageBubble`): the user's is trailing on a quiet fill, the agent's
 * leading plain prose (no bubble) rendered as Markdown, a system note footnote-secondary. [meta]
 * (time, delivery) sits under it, accent while [pending].
 */
@Composable
fun MessageBubble(
    role: MessageRole,
    text: String,
    modifier: Modifier = Modifier,
    meta: String? = null,
    pending: Boolean = false,
    maxBubbleWidth: Dp = 320.dp,
) {
    val colors = OptioTheme.colors
    val type = OptioTheme.type
    val alignment = if (role == MessageRole.USER) Alignment.End else Alignment.Start
    Column(modifier.fillMaxWidth(), horizontalAlignment = alignment, verticalArrangement = Arrangement.spacedBy(3.dp)) {
        when (role) {
            MessageRole.USER -> SelectionContainer {
                Text(
                    text,
                    style = type.body,
                    color = colors.label,
                    modifier = Modifier
                        .widthIn(max = maxBubbleWidth)
                        .background(colors.fillSecondary, Radius.bubbleShape)
                        .padding(horizontal = Spacing.m, vertical = Spacing.s),
                )
            }
            MessageRole.AGENT -> SelectionContainer(Modifier.fillMaxWidth()) {
                MarkdownText(text)
            }
            MessageRole.SYSTEM -> Text(text, style = type.footnote, color = colors.secondaryLabel)
        }
        if (meta != null) {
            Text(meta, style = type.caption2, color = if (pending) colors.accent else colors.tertiaryLabel)
        }
    }
}

/**
 * Monochrome pipeline strip (iOS `PipelineStrip`, the review detail's stages): done steps in the
 * label colour, the [current] one accent (red when [failed]), future ones quaternary.
 */
@Composable
fun PipelineStrip(
    steps: List<String>,
    current: Int,
    modifier: Modifier = Modifier,
    failed: Boolean = false,
) {
    val colors = OptioTheme.colors
    Row(modifier.fillMaxWidth(), horizontalArrangement = Arrangement.spacedBy(Spacing.xs)) {
        steps.forEachIndexed { index, step ->
            val fill = when {
                index < current -> colors.label
                index == current -> if (failed) colors.red else colors.accent
                else -> colors.quaternaryLabel
            }
            Column(Modifier.weight(1f), verticalArrangement = Arrangement.spacedBy(Spacing.xs)) {
                Box(Modifier.fillMaxWidth().height(3.dp).background(fill, Radius.capsuleShape))
                Text(
                    step,
                    style = OptioTheme.type.caption2,
                    color = if (index <= current) colors.label else colors.tertiaryLabel,
                    maxLines = 1,
                    overflow = TextOverflow.Ellipsis,
                )
            }
        }
    }
}
